File: py-claw/src/py_claw/context/skill.py
"""技能加载 — 从工作区 .claw/skills/ 目录加载 SKILL.md。"""

from __future__ import annotations

from pathlib import Path


class Skill:
    name: str
    description: str
    body: str


class SkillLoader:
    """加载工作区下的技能外挂文件。"""

    def __init__(self, work_dir: str) -> None:
        self.work_dir = work_dir

    def load_all(self) -> str:
        skill_dir = Path(self.work_dir) / ".claw" / "skills"
        if not skill_dir.is_dir():
            return ""

        lines: list[str] = []
        lines.append("\n### 可用专业技能 (Agent Skills)")
        lines.append("以下是你拥有的标准化外挂技能，请在符合描述的场景下严格遵循其正文指令：\n")

        for skill_file in sorted(skill_dir.rglob("SKILL.md")):
            content = skill_file.read_text(encoding="utf-8")
            skill = self._parse_skill_md(content)
            lines.append(f"#### 技能名称: {skill.name}")
            lines.append(f"**触发条件**: {skill.description}\n")
            lines.append("**执行指南**:")
            lines.append(skill.body)
            lines.append("\n---\n")

        result = "\n".join(lines)
        if len(result) < 50:
            return ""
        return result

    def _parse_skill_md(self, content: str) -> Skill:
        skill = Skill()
        skill.name = "Unknown Skill"
        skill.description = "No description provided."
        skill.body = content

        stripped = content.lstrip()
        if stripped.startswith("---"):
            parts = stripped.split("---", 2)
            if len(parts) == 3:
                skill.body = parts[2].strip()
                for line in parts[1].strip().splitlines():
                    line = line.strip()
                    if line.startswith("name:"):
                        skill.name = line[len("name:"):].strip()
                    elif line.startswith("description:"):
                        skill.description = line[len("description:"):].strip()

        return skill

File: ts-claw/src/context/skill.ts
/** 技能加载 — 从工作区 .claw/skills/ 目录加载 SKILL.md。 */

import { readdirSync, readFileSync, statSync } from 'node:fs';
import { join } from 'node:path';

export interface Skill {
  name: string;
  description: string;
  body: string;
}

export class SkillLoader {
  constructor(private workDir: string) {}

  loadAll(): string {
    const skillBase = join(this.workDir, '.claw', 'skills');
    try {
      if (!statSync(skillBase).isDirectory()) return '';
    } catch {
      return '';
    }

    const lines: string[] = [];
    lines.push('\n### 可用专业技能 (Agent Skills)');
    lines.push('以下是你拥有的标准化外挂技能，请在符合描述的场景下严格遵循其正文指令：\n');

    this._walkSkills(skillBase, lines);

    const result = lines.join('\n');
    return result.length < 50 ? '' : result;
  }

  private _walkSkills(dir: string, lines: string[]): void {
    let entries: import('node:fs').Dirent[];
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      const full = join(dir, entry.name);
      if (entry.isDirectory()) {
        this._walkSkills(full, lines);
      } else if (entry.name === 'SKILL.md') {
        const content = readFileSync(full, 'utf-8');
        const skill = this._parseSkillMd(content);
        lines.push(`#### 技能名称: ${skill.name}`);
        lines.push(`**触发条件**: ${skill.description}\n`);
        lines.push('**执行指南**:');
        lines.push(skill.body);
        lines.push('\n---\n');
      }
    }
  }

  private _parseSkillMd(content: string): Skill {
    const skill: Skill = { name: 'Unknown Skill', description: 'No description provided.', body: content };
    const trimmed = content.trimStart();
    if (trimmed.startsWith('---')) {
      const parts = trimmed.split('---', 3);
      if (parts.length === 3) {
        skill.body = parts[2].trim();
        for (const line of parts[1].trim().split('\n')) {
          const l = line.trim();
          if (l.startsWith('name:')) skill.name = l.slice(5).trim();
          else if (l.startsWith('description:')) skill.description = l.slice(12).trim();
        }
      }
    }
    return skill;
  }
}
